`TestReadyAndHealthy` CI test flaky

### What did you do?

-

### What did you expect to see?

_No response_

### What did you see instead? Under which circumstances?

The `TestReadyAndHealthy` test failed in CI, seems flaky:

```
--- FAIL: TestReadyAndHealthy (0.02s)
panic: Unable to start web listeners: listen tcp :38639: bind: address already in use [recovered]
	panic: Unable to start web listeners: listen tcp :38639: bind: address already in use

goroutine 133 [running]:
testing.tRunner.func1.2({0x12cf780, 0xc000620b90})
	/usr/local/go/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1635 +0x35e
panic({0x12cf780?, 0xc000620b90?})
	/usr/local/go/src/runtime/panic.go:785 +0x132
github.com/prometheus/prometheus/web.TestReadyAndHealthy(0xc000032b60)
	/__w/prometheus/prometheus/web/web_test.go:106 +0x108b
testing.tRunner(0xc000032b60, 0x159b7b0)
	/usr/local/go/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1743 +0x390
FAIL	github.com/prometheus/prometheus/web	10.078s
```

### System information

_No response_

### Prometheus version

_No response_

### Prometheus configuration file

_No response_

### Alertmanager version

_No response_

### Alertmanager configuration file

_No response_

### Logs

_No response_

--- a/util/testutil/port.go
+++ b/util/testutil/port.go
@@ -15,21 +15,56 @@ package testutil
 
 import (
 	"net"
+	"sync"
 	"testing"
+)
+
+var (
+	mu        sync.Mutex
+	usedPorts []int
 )
 
 // RandomUnprivilegedPort returns valid unprivileged random port number which can be used for testing.
 func RandomUnprivilegedPort(t *testing.T) int {
 	t.Helper()
+	mu.Lock()
+	defer mu.Unlock()
 
+	port, err := getPort()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for portWasUsed(port) {
+		port, err = getPort()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	usedPorts = append(usedPorts, port)
+
+	return port
+}
+
+func portWasUsed(port int) bool {
+	for _, usedPort := range usedPorts {
+		if port == usedPort {
+			return true
+		}
+	}
+	return false
+}
+
+func getPort() (int, error) {
 	listener, err := net.Listen("tcp", ":0")
 	if err != nil {
-		t.Fatalf("Listening on random port: %v", err)
+		return 0, err
 	}
 
 	if err := listener.Close(); err != nil {
-		t.Fatalf("Closing listener: %v", err)
+		return 0, err
 	}
 
-	return listener.Addr().(*net.TCPAddr).Port
+	return listener.Addr().(*net.TCPAddr).Port, nil
 }

